File: sample-chat-java/app/src/main/java/com/quickblox/sample/chat/java/utils/qb/QbDialogUtils.java
package com.quickblox.sample.chat.java.utils.qb;

import android.text.TextUtils;
import android.util.Log;

import com.quickblox.chat.model.QBChatDialog;
import com.quickblox.chat.model.QBDialogType;
import com.quickblox.chat.utils.DialogUtils;
import com.quickblox.sample.chat.java.utils.chat.ChatHelper;
import com.quickblox.users.model.QBUser;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class QbDialogUtils {
    private static final String TAG = QbDialogUtils.class.getSimpleName();


    public static QBChatDialog createDialog(List<QBUser> users, String chatName) {
        if (isPrivateChat(users)) {
            QBUser currentUser = ChatHelper.getCurrentUser();
            users.remove(currentUser);
        }
        QBChatDialog dialog = DialogUtils.buildDialog(users.toArray(new QBUser[users.size()]));
        if (!TextUtils.isEmpty(chatName)) {
            dialog.setName(chatName);
        }
        return dialog;
    }

    private static boolean isPrivateChat(List<QBUser> users) {
        return users.size() == 2;
    }

    public static List<QBUser> getAddedUsers(QBChatDialog dialog, List<QBUser> currentUsers) {
        return getAddedUsers(getQbUsersFromQbDialog(dialog), currentUsers);
    }

    public static List<QBUser> getAddedUsers(List<QBUser> previousUsers, List<QBUser> currentUsers) {
        List<QBUser> addedUsers = new ArrayList<>();
        for (QBUser currentUser : currentUsers) {
            boolean wasInChatBefore = false;
            for (QBUser previousUser : previousUsers) {
                if (currentUser.getId().equals(previousUser.getId())) {
                    wasInChatBefore = true;
                    break;
                }
            }
            if (!wasInChatBefore) {
                addedUsers.add(currentUser);
            }
        }

        QBUser currentUser = ChatHelper.getCurrentUser();
        addedUsers.remove(currentUser);

        return addedUsers;
    }

    public static List<QBUser> getRemovedUsers(QBChatDialog dialog, List<QBUser> currentUsers) {
        return getRemovedUsers(getQbUsersFromQbDialog(dialog), currentUsers);
    }

    public static List<QBUser> getRemovedUsers(List<QBUser> previousUsers, List<QBUser> currentUsers) {
        List<QBUser> removedUsers = new ArrayList<>();
        for (QBUser previousUser : previousUsers) {
            boolean isUserStillPresented = false;
            for (QBUser currentUser : currentUsers) {
                if (previousUser.getId().equals(currentUser.getId())) {
                    isUserStillPresented = true;
                    break;
                }
            }
            if (!isUserStillPresented) {
                removedUsers.add(previousUser);
            }
        }

        QBUser currentUser = ChatHelper.getCurrentUser();
        removedUsers.remove(currentUser);

        return removedUsers;
    }

    public static void logDialogUsers(QBChatDialog qbDialog) {
        Log.v(TAG, "Dialog " + getDialogName(qbDialog));
        logUsersByIds(qbDialog.getOccupants());
    }

    public static void logUsers(List<QBUser> users) {
        for (QBUser user : users) {
            Log.i(TAG, user.getId() + " " + user.getFullName());
        }
    }

    private static void logUsersByIds(List<Integer> users) {
        for (Integer id : users) {
            QBUser user = QbUsersHolder.getInstance().getUserById(id);
            Log.i(TAG, ((user != null && user.getId() != null) ? (user.getId() + " " + user.getLogin()) : "noId"));
        }
    }

    public static String getDialogName(QBChatDialog dialog) {
        if (dialog.getType().equals(QBDialogType.GROUP)) {
            return dialog.getName();
        } else {
            // It's a private dialog, let's use opponent's name as chat name
            Integer opponentId = dialog.getRecipientId();
            QBUser user = QbUsersHolder.getInstance().getUserById(opponentId);
            if (user != null) {
                return TextUtils.isEmpty(user.getFullName()) ? user.getLogin() : user.getFullName();
            } else {
                return dialog.getName();
            }
        }
    }

    private static List<QBUser> getQbUsersFromQbDialog(QBChatDialog dialog) {
        List<QBUser> previousDialogUsers = new ArrayList<>();
        for (Integer id : dialog.getOccupants()) {
            QBUser user = QbUsersHolder.getInstance().getUserById(id);
            if (user != null) {
                previousDialogUsers.add(user);
            }
        }
        return previousDialogUsers;
    }

    public static List<Integer> getOccupantsIdsListFromString(String occupantIds) {
        List<Integer> occupantIdsList = new ArrayList<>();
        if (occupantIds != null) {
            String[] occupantIdsArray = occupantIds.split(",");
            for (String occupantId : occupantIdsArray) {
                occupantIdsList.add(Integer.valueOf(occupantId.trim()));
            }
        }
        return occupantIdsList;
    }

    public static String getOccupantsIdsStringFromList(Collection<Integer> occupantIdsList) {
        return TextUtils.join(",", occupantIdsList);
    }

    public static String getOccupantsNamesStringFromList(Collection<QBUser> qbUsers) {
        ArrayList<String> userNameList = new ArrayList<>();
        for (QBUser user : qbUsers) {
            if (TextUtils.isEmpty(user.getFullName())) {
                userNameList.add(user.getLogin());
            } else {
                userNameList.add(user.getFullName());
            }
        }
        return TextUtils.join(", ", userNameList);
    }

    public static QBChatDialog buildPrivateChatDialog(String dialogId, Integer recipientId) {
        QBChatDialog chatDialog = DialogUtils.buildPrivateDialog(recipientId);
        chatDialog.setDialogId(dialogId);

        return chatDialog;
    }
}

File: sample-chat-kotlin/src/main/java/com/quickblox/sample/chat/kotlin/utils/qb/QbDialogUtils.kt
package com.quickblox.sample.chat.kotlin.utils.qb

import android.text.TextUtils
import android.util.Log
import com.quickblox.chat.model.QBChatDialog
import com.quickblox.chat.model.QBDialogType
import com.quickblox.chat.utils.DialogUtils
import com.quickblox.sample.chat.kotlin.utils.chat.ChatHelper
import com.quickblox.users.model.QBUser
import java.util.*
import kotlin.collections.ArrayList

private const val PRIVATE_CHAT_COUNT_USERS = 2
private const val TAG = "QbDialogUtils::class"

fun createDialog(users: MutableList<QBUser>, chatName: String): QBChatDialog {
    if (isPrivateChat(users)) {
        users.remove(ChatHelper.getCurrentUser())
    }
    val dialog = DialogUtils.buildDialog(*users.toTypedArray())
    if (!TextUtils.isEmpty(chatName)) {
        dialog.name = chatName
    }
    return dialog
}

private fun isPrivateChat(users: List<QBUser>): Boolean {
    return users.size == PRIVATE_CHAT_COUNT_USERS
}

fun getAddedUsers(dialog: QBChatDialog, currentUsers: List<QBUser>): List<QBUser> {
    return getAddedUsers(getQbUsersFromQbDialog(dialog), currentUsers)
}

private fun getAddedUsers(previousUsersList: List<QBUser>, currentUsersList: List<QBUser>): List<QBUser> {
    val addedUsers = ArrayList<QBUser>()
    for (currentUser in currentUsersList) {
        var wasInChatBefore = false
        for (previousUser in previousUsersList) {
            if (currentUser.id == previousUser.id) {
                wasInChatBefore = true
                break
            }
        }
        if (!wasInChatBefore) {
            addedUsers.add(currentUser)
        }
    }
    addedUsers.remove(ChatHelper.getCurrentUser())
    return addedUsers
}

fun getRemovedUsers(dialog: QBChatDialog, currentUsers: List<QBUser>): List<QBUser> {
    return getRemovedUsers(getQbUsersFromQbDialog(dialog), currentUsers)
}

private fun getRemovedUsers(previousUsersList: List<QBUser>, currentUsersList: List<QBUser>): List<QBUser> {
    val removedUsers = ArrayList<QBUser>()
    for (previousUser in previousUsersList) {
        var isUserStillPresented = false
        for (currentUser in currentUsersList) {
            if (previousUser.id == currentUser.id) {
                isUserStillPresented = true
                break
            }
        }
        if (!isUserStillPresented) {
            removedUsers.add(previousUser)
        }
    }
    removedUsers.remove(ChatHelper.getCurrentUser())
    return removedUsers
}

fun logDialogUsers(qbDialog: QBChatDialog) {
    Log.v(TAG, "Dialog " + getDialogName(qbDialog))
    logUsersByIds(qbDialog.occupants)
}

fun logUsers(users: List<QBUser>) {
    for (user in users) {
        Log.i(TAG, user.id.toString() + " " + user.fullName)
    }
}

private fun logUsersByIds(usersIds: List<Int>) {
    for (id in usersIds) {
        val user = QbUsersHolder.getUserById(id)
        Log.i(TAG, user?.id.toString() + " " + user?.fullName)
    }
}

fun getDialogName(dialog: QBChatDialog): String {
    var result = ""
    if (dialog.type == QBDialogType.GROUP) {
        result = dialog.name
    } else {
        // It's a private dialog, let's use opponent's name as chat name
        val opponentId = dialog.recipientId
        val user = QbUsersHolder.getUserById(opponentId)

        user?.let {
            result = if (TextUtils.isEmpty(it.fullName)) {
                it.login
            } else {
                it.fullName
            }
        } ?: run {
            result = dialog.name
        }
    }
    return result
}

private fun getQbUsersFromQbDialog(dialog: QBChatDialog): List<QBUser> {
    val previousDialogUsers = ArrayList<QBUser>()
    for (id in dialog.occupants) {
        val user = QbUsersHolder.getUserById(id)
        user?.let {
            previousDialogUsers.add(it)
        }
    }
    return previousDialogUsers
}

fun getOccupantsIdsListFromString(occupantIds: String?): List<Int> {
    val occupantIdsList = ArrayList<Int>()
    occupantIds?.let {
        val occupantIdsArray = occupantIds.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (occupantId in occupantIdsArray) {
            occupantIdsList.add(Integer.valueOf(occupantId.trim()))
        }
    }
    return occupantIdsList
}

fun getOccupantsIdsStringFromList(occupantIdsList: Collection<Int>): String {
    return TextUtils.join(",", occupantIdsList)
}

fun getOccupantsNamesStringFromList(qbUsers: Collection<QBUser>): String {
    val userNameList = ArrayList<String>()
    for (user in qbUsers) {
        if (TextUtils.isEmpty(user.fullName)) {
            userNameList.add(user.login)
        } else {
            userNameList.add(user.fullName)
        }
    }
    return TextUtils.join(", ", userNameList)
}